frontend/lib/resolution/copy-elision.cpp

```diff
@@ -404,113 +406,83 @@ void FindElidedCopies::handleYield(const uast::Yield* ast, RV& rv) {
-          const CopyElisionState& elseState = elseFrame->copyElisionState[id];
-          if (elseState.lastIsCopy) {
-            CopyElisionState& s = frame->copyElisionState[id];
-            s.lastIsCopy = true;
-            s.points.insert(thenState.points.begin(), thenState.points.end());
-            s.points.insert(elseState.points.begin(), elseState.points.end());
-          }
-        }
-      }
-    }
-  } else {
-    // Both then and else exist, but one block returned and the
-    // other did not. Promote all elision points for the block that
-    // didn't return into the parent map.
-    // (those from the block that returned have already been saved)
-    VarFrame* branch = thenReturnsThrows ? elseFrame : thenFrame;
-
-    // First, save local variables from the block
-    saveLocalVarElidedCopies(branch);
-
-    // promote other vars in the branch not returning
-    for (const auto& pair : branch->copyElisionState) {
-      const CopyElisionState& branchState = pair.second;
-      if (branchState.lastIsCopy) {
-        ID id = pair.first;
-        // only consider variables that aren't local to the branch
-        if (branch->eligibleVars.count(id) == 0) {
-          CopyElisionState& s = frame->copyElisionState[id];
-          s.lastIsCopy = true;
-          s.points.insert(branchState.points.begin(), branchState.points.end());
-        }
+
+void FindElidedCopies::handleDisjunction(const AstNode * node,
+                                         VarFrame * currentFrame, 
+                                         const std::vector<VarFrame*>& frames, 
+                                         bool total, RV& rv) {
+  // if any frame is 'param true' the rest of the frames do not matter
+  for (auto frame : frames) {
+    if (!frame->knownPath) continue;
+    // yielding handled during return/throw
+    if (frame->returnsOrThrows) return;
+
+    //yield the local variables
+    saveLocalVarElidedCopies(frame);
+
+    //promote outer varaible access
+    for (const auto& entry : frame->copyElisionState) {
+      ID id = entry.first;
+      const CopyElisionState& state = entry.second;
+      if (state.lastIsCopy && frame->eligibleVars.count(id) == 0) {
+        CopyElisionState& s = currentFrame->copyElisionState[id];
+        s.lastIsCopy = true;
+        s.points.insert(state.points.begin(), state.points.end());
       }
     }
+
+    propogateMentionsAndInits(currentFrame, frame);
+
+    return;
   }
 
-  // propagate inited variables from the then/else scopes
-  if (thenFrame) {
-    frame->initedVars.insert(thenFrame->initedVars.begin(),
-                             thenFrame->initedVars.end());
+  //propogate mentions to the parent
+  for(auto frame: frames) {
+    propogateMentionsAndInits(currentFrame, frame);
   }
-  if (elseFrame) {
-    frame->initedVars.insert(elseFrame->initedVars.begin(),
-                             elseFrame->initedVars.end());
+
+  if (!total) return;
```

Comment:
Are you sure you can do this? Even if it's non-total and one of the branches has no elision stuff, I think you need to call `saveLocalVarElidedCopies` on each (valid) frame, since local copies are safe to yield.